### PR TITLE
Make feedback messages associated to a canonical state_id [Under Revision]

### PR DESCRIPTION
**Pull request by ejgallego**
This is the first of a series of pull requests trying to streamline the
construction of full documents using the Stm by IDE/SerAPI.

Currently, `feedback` carries an `edit_or_state_id`, but the `edit_id`
case is used once (to report parsing errors) and indeed that case is
handled specifically in CoqIDE and PIDEtop by catching the exception.

For IDEs and SerAPI, it makes more sense to always map feedback messages
to already "added" parts of the document, simplifying message handling,
and treat document building failure as a different case.

This patch removes `edit_id` from feedback messages. Tested CoqIDE,
seems unaffected, PIDE should work fine too (but I lack eclipse).

No changes to the protocol format. No more `ErrorMsg` will be generated
on parse error, consumers can still emit a custom error by catching the
exception (as they do now) or by hooking into `Stm.parse_error_hook`

### CODE REVIEW

File: stm/stm.ml [hunk 1]
```diff
@@ -114,10 +113,12 @@ let vernac_interp ?proof id ?route { verbose; loc; expr } =
...
-let vernac_parse ?newtip ?route eid s =
-  let feedback_id = 
-    if Option.is_empty newtip then Edit eid
-    else State (Option.get newtip) in
+
+(* We report feedback for parsing errors with the stateId the parsing
+   is being attempted on. This is more convenient for IDEs, as the can
+   manage the state in an easier way only caring about stateIds.
+*)
+let vernac_parse ?route feedback_id eid s =
```
Comment (ejgallego) [inline stm/stm.ml]: In fact we could remove the feedback use here altogether, what useful feedback can a non-parseable statement produce?

**Comment by gares** [comment]
The patch is wrong.  If you have "(1) Foo. (2)  Bar. (3)", when you add "Bar" the cur_tip you use to give feedback on Bar is the state corresponding to the execution of Foo, 2 here, while it should be (3).  Currently a parse error is reported on the edit_id of "Bar".

What one could do is to make the ?new_tip argument mandatory, and feedback on that.  That should work.  Unfortunately there is no garantee the new_tip you pass will be the same the STM gives in output (corner cases with nested stuff).  E.g. PIDE passes new_tip, but also deals with that case... ugly.

Wrt your question about feedback on the currently parsed sentence, well some feedback may occur before the parsing error is reached.  Today the parser emits no feedback IIRC, but could.  E.g. in Isabelle it is the parser that tells the UI is "foo" is a keyword or not, so that syntax highlighting is not hard coded in the UI.

Last edit_it had originally another purpose in the Isabelle model:  the UI chops the document into sentences, then "internalizes" them assigning to each sentence an edit_id, then declares a document by simply listing the ids.  This for performance reasons.  Coq is unable today to take advantage of that way of doing things, so I'm fine trying to get rid of it, but I'd love a design that does not necessarily require an API change if we decide the STM does not re-parse a sentence.

So, in short, this patch is not OK, but I don't have a clean solution to get rid of the ugly edit_id.

**Comment by ejgallego** [comment]
Hi Enrico, thanks again for the quick review!

I agree with your comment, the patch should be tweaked and renamed to "Forbid feedback at parsing time".

Not that I'm proud of the path taken here, but I consider it a bit better than the current situation. I'll think a more about this, a few comments/quesitons for now:

- What does Isabelle do at parsing error time? Does it still assign an `edit_id`.
- Doesn't `Stm.add` take the role of the "document splitter" already in Coq? At least, as you can see from #204 this is what we are trying to do. Given your description `edit_id` seems to exactly correspond to `Stateid.t` in Coq.
- How are clients supposed to make use of `edit_id` feedback? So far, no client seems to use it.

  Indeed, this is the main issue and my main motivation for writing this patch. So far, the only way for clients to reliably determine the end of a sentence is to call `Stm.add` an use the returned location. Thus feedback coming with an `edit_id` is basically ignored. An indeed, what is the point them? It seems much more convenient to IDEs to map each piece of feedback to a _valid_ piece of Coq document.

  I see what you say about the parser giving feedback, but IMO with the current design, a whole new message type would be needed.

- BTW: I thought that in the current model, removing `new_tip` altogether would make sense, however IDE writers like to assign themselves the ID. Dunno about this.

**Comment by gares** [comment]
edit_id are called span_ids in Isabelle and assigned by the UI to sentences.  Then a message says "declare `some text` as span_id -356"... later on a message may say "The current document is composed of span_id -36, -456, -356"... the isabelle replies "feedback for sentence -36 are on execution_id [99;100], for -456 on execution_id [100], ..." where execution_id is a unique integer assigned to the (execution of a) sentence by Isabelle (called state_id in Coq).
Then any feedback on sentence -36 is either on -36 or on 99 or on 100.  Clearly the interning of spans requires a notion of span_id, something that CoqIDE does not do and hence is not in STM (it is in pidetop for example, that speaks isabelle's protocol).  So these edit_ids seem useless to you but they still play a role.

A more general views is: any sentence (identified by Coq or by the UI, it does not really matter) has attached to it a number of unique ids.  Some chosen by the UI, some by Coq.  Feedback in on any of these.
Note that allowing multiple ids one can do smart stuff (not in CoqIDE, but Isabelle's protocol is open to that).  If one adds to the document a sentence already added before, and the STM has a cache, it could assign the same execution_id (to the same span_id) and the UI could just keep the feedback already received.  If STM assigned to it a different execution_id then the UI drops all feedback received so far, since the span is being re-evaluated.

Maybe another way to explain why edit_ids are needed is the following one.  Imagine ou are fully asynchronous, you send to Coq the entire text over and over and Coq sends back feedback.  You send text1 and then text2 before receiving any feedback.  There is a parse error.  In order to signal it on text1 or text2 Coq needs to attach to the two texts different ids.  what does the UI do? use the order of the messages?  No, it is much simple to say (text1,eid1)... (text2,eid2)... and then have Coq mention eid1 or eid2 in its answers.  The protocol of CoqIDE waits for a (eid1,sid1) before going on, and indeed document communication is synchronous.  So one could get rid of eid1.  But if you don't wait for the answer the eid is really needed to make sense of Coq's answer.

**Comment by ejgallego** [comment]
Thanks a lot Enrico for your detailed message. I need some more time to think, for now let's push this to 8.7.

**Comment by ejgallego** [comment]
Reporting at add time is broken as `edit_id` is not used once the classifier forces observation. See: https://coq.inria.fr/bugs/show_bug.cgi?id=4972

**Comment by ejgallego** [comment]
Superseded by #441
